nixos/modules/programs/cpu-energy-meter.nix

```
@@ -0,0 +1,27 @@
+{ config
+, lib
+, pkgs
+, ...
+}: {
+  options.programs.cpu-energy-meter = {
+    enable = lib.mkEnableOption "CPU Energy Meter";
+    package = lib.mkPackageOption pkgs "cpu-energy-meter" { };
+  };
+
+  config =
+    let
+      cfg = config.programs.cpu-energy-meter;
+    in
+    lib.mkIf cfg.enable {
+      hardware.cpu.x86.msr.enable = true;
+
```

```suggestion
      boot.kernelModules = [ "cpuid" ];
```
The [readme](https://github.com/sosy-lab/cpu-energy-meter) suggests that cpuid should be loaded I believe.

---

It does, but on my machine I could not verify that it is required:

```
$ sudo lsmod | grep -E "^(msr|cpuid)"
msr                    16384  0
$ cpu-energy-meter -d
[DEBUG] GenuineIntel processor found.
[DEBUG] Processor is from family 6 and uses model 0xA0650.
[DEBUG] Using /dev/cpu/0/msr for accessing MSR of socket 0.
[DEBUG] Domain Package is supported.
[DEBUG] Domain Core is supported.
[DEBUG] Domain Uncore is supported.
[DEBUG] Domain DRAM is supported.
[DEBUG] Domain PSYS is supported.
[DEBUG] Measured the following unit multipliers:   RAPL_ENERGY_UNIT=6.103516e-05J   RAPL_DRAM_ENERGY_UNIT=6.103516e-05J
[DEBUG] Not changing UID because not running as root (uid=1000 gid=100).
[DEBUG] Max power consumption of node 0 is 125.000000W.
[DEBUG] Interval time of msr probes set to 1047s, 0ns.
^C
[DEBUG] Received signal 2.
+--------------------------------------+
| CPU Energy Meter            Socket 0 |
+--------------------------------------+
Duration                  7.913466 sec
Package                 169.754578 Joule
Core                    163.473511 Joule
DRAM                     11.148376 Joule
$ sudo modprobe cpuid
$ sudo lsmod | grep -E "^(msr|cpuid)"
cpuid                  16384  0
msr                    16384  0
$ gunzip --stdout /proc/config.gz | grep -s 'CONFIG_X86_CPUID'
CONFIG_X86_CPUID=m
```

@PhilippWendler is loading `cpuid` stricly required?

---

Good question! I think I kept this requirement from the code base from Intel on which cpu-energy-meter was originally based. I don't even know exactly what the cpuid module does. [`cpuid(4)`](https://man7.org/linux/man-pages/man4/cpuid.4.html says it provides `/dev/cpu/*/cpuid`, but we do not use this, we execute the `cpuid` instruction. So as long as there is no implicit requirement somewhere hidden (which I don't think there is), the module should not be necessary.

I will update our readme accordingly.